pallets/funding/src/functions/misc.rs

```diff
@@ -413,9 +421,168 @@ impl<T: Config> Pallet<T> {
+	}
+
+	pub fn verify_receiving_account_signature(
+		polimec_account: &AccountIdOf<T>,
+		project_id: ProjectId,
+		receiver_account: &Junction,
+		signature_bytes: [u8; 65],
+	) -> DispatchResult {
+		match receiver_account {
+			Junction::AccountId32 { network, id } =>
+				if network.is_none() {
+					let signature = SrSignature::from_slice(&signature_bytes[..64])
+						.map_err(|_| Error::<T>::BadReceiverAccountSignature)?;
+					let public = SrPublic::from_slice(id).map_err(|_| Error::<T>::BadReceiverAccountSignature)?;
+					ensure!(
+						signature.verify(message_bytes.as_slice(), &public),
+						Error::<T>::BadReceiverAccountSignature
+					);
+				},
+
+			Junction::AccountKey20 { key: expected_ethereum_account, .. } => {
+				ensure!(
+					Self::verify_ethereum_account(
+						signature_bytes,
+						*expected_ethereum_account,
+						polimec_account.clone(),
+						project_id,
+					),
+					Error::<T>::BadReceiverAccountSignature
+				);
+			},
+			_ => return Err(Error::<T>::UnsupportedReceiverAccountJunction.into()),
+		};
+		Ok(())
+	}
+
 	pub fn get_decimals_aware_funding_asset_price(funding_asset: &AcceptedFundingAsset) -> Option<PriceOf<T>> {
 		let funding_asset_id = funding_asset.id();
 		let funding_asset_decimals = T::FundingCurrency::decimals(funding_asset_id);
 		<PriceProviderOf<T>>::get_decimals_aware_price(funding_asset_id, USD_DECIMALS, funding_asset_decimals)
 	}
 }
+
+pub mod typed_data_v4 {
+	use super::*;
+
+	/// Returns the first part needed for a typed data v4 message. It specifies the entity that requires the signature.
+	pub fn get_domain_separator(name: &str, version: &str, chain_id: u32, verifying_contract: &str) -> [u8; 32] {
+		let encoded_domain_type =
+			keccak_256(b"EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
+		let encoded_name = keccak_256(name.as_bytes());
+		let encoded_version = keccak_256(version.as_bytes());
+
+		// TODO: Handle panics from copy_from_slice
+		// u32 should be converted to u256 in big endian notation
+		let chain_id_bytes: [u8; 4] = chain_id.to_be_bytes();
+		let mut encoded_chain_id: [u8; 32] = [0u8; 32];
+		encoded_chain_id[28..].copy_from_slice(&chain_id_bytes);
+		// Should be 32 bytes to comply with ABI encoding, so we pad with zeroes to the left
+		let mut encoded_contract: [u8; 32] = [0u8; 32];
+		encoded_contract[12..].copy_from_slice(hex::decode(verifying_contract).unwrap().as_slice());
+
+		let mut data = Vec::new();
+		data.extend_from_slice(&encoded_domain_type);
+		data.extend_from_slice(&encoded_name);
+		data.extend_from_slice(&encoded_version);
+		data.extend_from_slice(&encoded_chain_id);
+		data.extend_from_slice(&encoded_contract);
+
+		keccak_256(&data)
+	}
+
+	/// Returns the second part needed for a typed data v4 message. It specifies the message details with type information.
+	pub fn get_message(polimec_account: &str, project_id: u32, nonce: u32) -> [u8; 32] {
```

Comment:
Similar to before, we can precompute the hash and pre-allocate a fixed size array for all components